src/Raven.Server/Commercial/LetsEncryptUtils.cs

```diff
@@ -0,0 +1,968 @@
+                }
+                catch (Exception e)
+                {
+                    if (cts.IsCancellationRequested == false)
+                        throw;
+                    throw new TimeoutException("Request failed due to a timeout error", e);
+                }
+            }
+        }
+
+        public static async Task<(string Challenge, LetsEncryptClient.CachedCertificateResult Cache)> InitialLetsEncryptChallenge(
+            SetupInfo setupInfo,
+            LetsEncryptClient client,
+            CancellationToken token)
+        {
+            try
+            {
+                var host = (setupInfo.Domain + "." + setupInfo.RootDomain).ToLowerInvariant();
+                var wildcardHost = "*." + host;
+                if (client.TryGetCachedCertificate(wildcardHost, out var certBytes))
+                    return (null, certBytes);
+
+                var result = await client.NewOrder(new[] {wildcardHost}, token);
+
+                result.TryGetValue(host, out var challenge);
+                // we may already be authorized for this?
+                return (challenge, null);
+            }
+            catch (Exception e)
+            {
+                throw new InvalidOperationException("Failed to receive challenge(s) information from Let's Encrypt.", e);
+            }
+        }
+
+        public class CompleteAuthorizationAndGetCertificateParameters
+        {
+            public Action OnValidationSuccessful;
+            public SetupProgressAndResult Progress;
+            public SetupInfo SetupInfo;
+            public LetsEncryptClient Client;
+            public (string Challange, LetsEncryptClient.CachedCertificateResult Cache) ChallengeResult;
+            public ServerStore ServerStore;
+            public CancellationToken Token;
+        }
+
+        public class CompleteClusterConfigurationParameters
+        {
+            public SetupProgressAndResult Progress;
+            public Action<IOperationProgress> OnProgress;
+            public SetupInfo SetupInfo;
+            public Func<string,string, Task> OnBeforeAddingNodesToCluster;
+            public Func<string, Task> AddNodeToCluster;
+            public Func<Action<IOperationProgress>, SetupProgressAndResult, X509Certificate2, Task> RegisterClientCertInOs;
+            public Func<StudioConfiguration.StudioEnvironment, Task> OnPutServerWideStudioConfigurationValues;
+            public Func<string, Task> OnWriteSettingsJsonLocally;
+            public Func<string, Task<string>> OnGetCertificatePath;
+            public Func<string, Task> RegisterClientCert;
+            public Func<CertificateDefinition, Task> PutCertificate;
+            public SetupMode SetupMode;
+            public string SettingsPath;
+            public bool CertificateValidationKeyUsages;
+            public LicenseType LicenseType;
+            public CancellationToken Token = CancellationToken.None;
+        }
```

Comment:
Wouldn't make sense to move nested classes to separate files? This class has almost 1K of lines